flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java

```diff
@@ -85,65 +86,68 @@ public static TypeInfo toHiveTypeInfo(DataType dataType) {
+						throw new CatalogException(
+								String.format("HiveCatalog doesn't support char type with length of '%d'. " +
+										"The maximum length is %d",
+										charType.getLength(), HiveChar.MAX_CHAR_LENGTH));
+					}
+					return TypeInfoFactory.getCharTypeInfo(charType.getLength());
 				}
-
-				return TypeInfoFactory.getCharTypeInfo(charType.getLength());
-			} else if (type.equals(LogicalTypeRoot.VARCHAR)) {
-				VarCharType varCharType = (VarCharType) dataType.getLogicalType();
-
-				// Flink's StringType is defined as VARCHAR(Integer.MAX_VALUE)
-				// We don't have more information in LogicalTypeRoot to distringuish StringType and a VARCHAR(Integer.MAX_VALUE) instance
-				// Thus always treat VARCHAR(Integer.MAX_VALUE) as StringType
-				if (varCharType.getLength() == Integer.MAX_VALUE) {
-					return TypeInfoFactory.stringTypeInfo;
+				case VARCHAR: {
+					VarCharType varCharType = (VarCharType) dataType.getLogicalType();
+					// Flink's StringType is defined as VARCHAR(Integer.MAX_VALUE)
+					// We don't have more information in LogicalTypeRoot to distringuish StringType and a VARCHAR(Integer.MAX_VALUE) instance
+					// Thus always treat VARCHAR(Integer.MAX_VALUE) as StringType
+					if (varCharType.getLength() == Integer.MAX_VALUE) {
+						return TypeInfoFactory.stringTypeInfo;
+					}
+					if (varCharType.getLength() > HiveVarchar.MAX_VARCHAR_LENGTH) {
+						throw new CatalogException(
+								String.format("HiveCatalog doesn't support varchar type with length of '%d'. " +
+										"The maximum length is %d",
+										varCharType.getLength(), HiveVarchar.MAX_VARCHAR_LENGTH));
+					}
+					return TypeInfoFactory.getVarcharTypeInfo(varCharType.getLength());
 				}
-
-				if (varCharType.getLength() > HiveVarchar.MAX_VARCHAR_LENGTH) {
-					throw new CatalogException(
-						String.format("HiveCatalog doesn't support varchar type with length of '%d'. " +
-								"The maximum length is %d",
-							varCharType.getLength(), HiveVarchar.MAX_VARCHAR_LENGTH));
+				case DECIMAL: {
+					DecimalType decimalType = (DecimalType) dataType.getLogicalType();
+					// Flink and Hive share the same precision and scale range
+					// Flink already validates the type so we don't need to validate again here
+					return TypeInfoFactory.getDecimalTypeInfo(decimalType.getPrecision(), decimalType.getScale());
 				}
-
-				return TypeInfoFactory.getVarcharTypeInfo(varCharType.getLength());
-			} else if (type.equals(LogicalTypeRoot.DECIMAL)) {
-				DecimalType decimalType = (DecimalType) dataType.getLogicalType();
-
-				// Flink and Hive share the same precision and scale range
-				// Flink already validates the type so we don't need to validate again here
-				return TypeInfoFactory.getDecimalTypeInfo(decimalType.getPrecision(), decimalType.getScale());
+				case VARBINARY: {
+					VarBinaryType varBinaryType = (VarBinaryType) dataType.getLogicalType();
+					if (varBinaryType.getLength() == VarBinaryType.MAX_LENGTH) {
```

Comment:
Nit: maybe we can add a comment saying the type is actually DataTypes.BYTES, similar to what is done for VARCHAR.

Reply:
That's an insightful opinion.